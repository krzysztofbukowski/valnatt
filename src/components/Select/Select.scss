@import '../../themes/ab';

.select {
  appearance: none;
  background-color: $select-background-color;
  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M7 10l5 5 5-5z" fill="#bfbfbf"/><path d="M0 0h24v24H0z" fill="none"/></svg>');
  background-repeat: no-repeat;
  background-position: calc(100% - 8px);
  border: 1px solid $border-color;
  box-shadow: 0 0 1px rgba(0,0,0, .1);
  height: 41px;
  font-size: inherit;
  font-weight: bold;
  outline: none;
  padding: 0 32px 0 16px;

  &:disabled {
    background-color: $select-background-color-disabled;
  }

  &.selected {
    background-color: $select-background-color-selected;
  }
}