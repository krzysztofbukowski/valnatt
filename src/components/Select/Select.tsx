import * as React from 'react';
import * as styles from './Select.scss';
import className from '../../utils/className';

export interface OptionGroup {
  label: string;
  options: { id: string, name: string }[];
}

export interface SelectProps {
  name: string;
  title?: string;
  options?: { id: string, name: string }[];
  optionGroups?: Array<OptionGroup>;
  current?: string;
  disabled?: boolean;
  onChange?: (value: string | null) => void;
}

interface SelectState {
  isSelected: boolean;
}

class Select extends React.Component<SelectProps, SelectState> {

  constructor(props: SelectProps) {
    super(props);

    this.state = {
      isSelected: !!props.current
    };
  }

  componentDidUpdate(prevProps: SelectProps) {
    if (this.props.current !== prevProps.current) {
      this.setState({
        isSelected: !!this.props.current
      });
    }
  }

  render() {
    const { options, name, current, title, optionGroups } = this.props;
    const hasOptions = this.hasOptions(options);
    const hasOptionGroups = this.hasOptionGroups(optionGroups);

    const classes = {
      [styles.select]: true,
      [styles.selected]: this.state.isSelected && hasOptions,
    };

    return (
      <select
        name={name}
        className={className(classes)}
        disabled={this.props.disabled}
        onChange={this.onChange}
        value={current || ''}
      >      
        {title && <option key={-1} value={null}>- {title} -</option>}
        {hasOptionGroups && this.renderOptionGroups(optionGroups)}
        {!hasOptionGroups && hasOptions && this.renderOptions(options)}
      </select>
    );
  }

  private hasOptions(options: { id: string, name: string }[]) {
    return options && options.length > 0;
  }

  private hasOptionGroups(options: Array<OptionGroup>) {
    return options && Object.keys(options).length > 0;
  }

  private renderOptions(options: { id: string, name: string }[]) {
    return options.map((option: { id: string, name: string }, optionIndex: number) => {
      return (
        <option key={optionIndex} value={option.id}>{option.name}</option>
      );
    });
  }

  private renderOptionGroups(optionGroups: Array<OptionGroup>) {
    return optionGroups &&
      optionGroups.map((optionGroup: OptionGroup, key: number) => {
        return (
          <optgroup
            key={key}
            label={optionGroup.label}>
            {this.renderOptions(optionGroup.options)}
          </optgroup>
        );
      });
  }

  private onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const isSelected = event.target.value !== `- ${this.props.title} -`;
    this.setState({
      isSelected: isSelected
    });

    if (this.props.onChange) {
      this.props.onChange(isSelected ? event.target.value : null);
    }
  }
}

export default Select;