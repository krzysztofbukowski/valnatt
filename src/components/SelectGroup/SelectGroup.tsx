import * as React from 'react';
import * as styles from './SelectGroup.scss';
import { SelectProps } from '../Select/Select';

export interface SelectGroupProps {
  name: string;
  onChange?: (value: string | null) => void;
}

export interface SelectGroupState {
  currentValue: string[];
}

class SelectGroup extends React.Component<SelectGroupProps, SelectGroupState> {
  constructor(props: SelectGroupProps) {
    super(props);

    let currentValue = new Array(React.Children.count(this.props.children));    

    React.Children.forEach(this.props.children, (child: React.ReactElement<SelectProps>, index: number) => {
      currentValue[index] = child.props.current;
    });

    this.state = {
      currentValue
    };
  }

  render() {
    const { children } = this.props;
    if (children === undefined) {
      throw new TypeError('SelectGroup component needs to have at least 2 children');
    }

    if (!(children as Object).hasOwnProperty('length')) {
      throw new TypeError('SelectGroup component needs to have at least 2 children');
    }

    return (
      <div className={styles.selectGroup}>
        {
          React.Children.map(this.props.children, this.mapChild)
        }
      </div>
    );
  }

  mapChild = (child: React.ReactElement<SelectProps>, key: number) => {
    const Component = child.type;
    const { options } = child.props;
    const { currentValue } = this.state;

    const props = {
      ...child.props,
      disabled: key > 0 && typeof (currentValue[key - 1]) !== 'string',
      options: options,
      current: currentValue[key],
      onChange: (value: string) => {
        this.onChange(key, value);
      }
    };

    return (<Component key={`select-${key}`} {...props} />);
  }

  onChange = (key: number, value: string | null) => {
    this.setState((prevState: SelectGroupState) => {
      const newValues = Array.from(prevState.currentValue);
      newValues[key] = value;

      for (let i = key + 1; i < newValues.length; i++) {
        newValues[i] = null;
      }

      return { ...prevState, currentValue: newValues };
    });

    let nextValue = value;
    if (nextValue === null && key > 0) {
      nextValue = this.state.currentValue[key - 1];
    }

    if (this.props.onChange) {
      this.props.onChange(nextValue);
    }
  }
}

export default SelectGroup;