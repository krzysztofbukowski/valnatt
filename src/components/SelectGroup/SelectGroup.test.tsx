import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import SelectGroup, { SelectGroupState } from './SelectGroup';
import Select from '../Select/Select';

describe('SelectGroup', () => {
  const countries = [
    {id: '1', name: 'Poland'},
    {id: '2', name: 'Sweden'},
    {id: '3', name: 'United Kingdom'},
  ];

  const cities = [
    {id: '11', name: 'Gdansk'},
    {id: '12', name: 'Warsaw'},
    {id: '13', name: 'Krakow'},
  ];

  it('should throw error without any children', () => {
    const wrapper = () => {
      shallow(<SelectGroup name="test" />);
    };

    expect(wrapper).toThrow(TypeError);
  });

  it('should throw error if only one child exists', () => {
    const wrapper = () => {
      shallow(
      <SelectGroup name="test" >
        <Select name="test" title="test" options={[]} />
      </SelectGroup>
      );
    };

    expect(wrapper).toThrow(TypeError);
  });

  it('should render correctly', () => {
    const tree = renderer.create(
      <SelectGroup name="area">
        <Select name="country" title="country" options={[]} />
        <Select name="city" title="city" options={[]} />
      </SelectGroup>
    ).toJSON();

    expect(tree).toMatchSnapshot();    
  });

  it('should have the default state correct', () => {
    [
      [undefined, undefined],
      ['1', undefined],
      ['1', '12']
    ].forEach((current: [string, string]) => {
      const component = shallow(
        <SelectGroup name="area">
          <Select name="country" title="country" options={countries} current={current[0]}/>
          <Select name="city" title="city" options={cities} current={current[1]}/>
        </SelectGroup>
      );
  
      const state = component.state() as SelectGroupState;
      expect(state.currentValue).toEqual(current);
    });
  });

  it('should fire change event with correct value', () => {
    const onChange = jest.fn();

    const component = mount(
      <SelectGroup name="area" onChange={onChange}>
        <Select name="country" title="country" options={countries}/>
        <Select name="city" title="city" options={cities}/>
      </SelectGroup>
    );

    component.find('select[name="country"]').simulate('change', { target: {value: '1'}});

    expect(onChange).toHaveBeenCalledWith('1');
  });

  it('should enable next select on change', () => {
    const component = mount(
      <SelectGroup name="area">
        <Select name="country" title="country" options={countries}/>
        <Select name="city" title="city" options={cities}/>
      </SelectGroup>
    );

    const citySelect = component.find(Select).last();
    expect(citySelect.props().disabled).toBeTruthy();

    component.find('select[name="country"]').simulate('change', {target: {value: '1'}});

    expect(component.find(Select).last().props().disabled).toBeFalsy();
  });

  it('should enable next select on change', () => {
    const component = mount(
      <SelectGroup name="area">
        <Select name="country" title="country" options={countries} current="1"/>
        <Select name="city" title="city" options={cities}/>
      </SelectGroup>
    );

    // const citySelect = component.find(Select).last();
    // expect(citySelect.props().disabled).toBeTruthy();

    // component.find('select[name="country"]').simulate('change', {target: {value: '1'}});

    // expect(component.find(Select).last().props().disabled).toBeFalsy();
  });   
});