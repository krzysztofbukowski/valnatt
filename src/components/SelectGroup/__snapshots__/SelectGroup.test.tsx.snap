// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectGroup should render correctly 1`] = `
<div>
  <select
    className=""
    disabled={false}
    name="country"
    onChange={[Function]}
    value=""
  >
    <option
      value={null}
    >
      - 
      country
       -
    </option>
  </select>
  <select
    className=""
    disabled={true}
    name="city"
    onChange={[Function]}
    value=""
  >
    <option
      value={null}
    >
      - 
      city
       -
    </option>
  </select>
</div>
`;
